_drafts/assert.md

```diff
@@ -0,0 +1,662 @@
+```
+
+The longer the string, the more codespace that is used, so this would actually
+**discourage** the use of asserts, which is not ideal.
+
+#### Extracted Information
+
+We can look up the file and line number of the assert hit assuming all assert
+strings are unique.
+
+### File**path** and Line Number Assert (#3)
+
+The most common approach I've seen for asserts is to include the filepath and
+line number into the assert macro, using the `__FILE__` and `__LINE__` defines
+provided by the compiler.
+
+```c
+#define MY_ASSERT(expr) \
+  do {                                          \
+    if (!(expr)) {                              \
+      my_assert(__FILE__, __LINE__);            \
+    }                                           \
+  } while (0)
+```
+
+```c
+void my_assert(const char *file, uint32_t line) {
+  strncpy(g_assert_info.file, file, sizeof(g_assert_info.file));
+  g_assert_info.line = line;
+  __asm("bkpt 3");
+}
+
+void assert_path_A(void) {
+  MY_ASSERT(0);
+}
+```
+
+This approach works very well and gives the developer exactly the amount of
+context required to find the issue. However, the `__FILE__` macro includes the
+**entire** filepath of the file, and those can get quite lengthy. In my local
+build, `__FILE__` is 77 bytes!
+
+```
+(gdb) p __FILE__
+$1 = "/Users/tyler/dev/memfault/interrupt/example/debugging-asserts/impls/assert3.c"
+```
+
+Thankfully, the `__FILE__` string table entry is reused between asserts in the
+same file, so the code size bloat primarily depends on the number of files with
+asserts within them. However, most projects have a non-trivial number of files.
+
+As a rough estimate, if we have asserts in 50 different files, this results in
+**~50 bytes per filepath \* 50 files => 2500 bytes**, which is an enormous
+amount of bloat just for storing filepaths!
+
+For this reason, many RTOS implementations that include an `ASSERT` macro can
+optionally leave these strings out of the build, such as Zephyr[^8] and
+MyNewt[^9].
+
+> NOTE: One simple code size (and sanity) optimization is to normalize these
+> `__FILE__` paths to the root of the project directory. This will result in a
+> build path that removes the `/Users/tyler/dev/memfault/...` portion. An added
+> benefit of this is that all all machines should produce builds with the same
+> filepaths, resulting in more similarly sized builds.
+
+#### Extracted Information
+
+We know the file and line number of the assert hit.
+
+### File**name** and Line Number Assert (#4)
+
+Since the filepaths can be lengthy and waste code space, one solution is to use
+only the filename itself, e.g. `assert3.c`, instead of the filepath, e.g.
+`/Users/tyler/.../assert3.c`.
+
+This can be accomplished by adding a define called `__FILENAME__` and using this
+in the assert implementation instead. In our Makefile or similar build system,
+this is defined along with the other `CFLAGS` as follows.
+
+```
+CFLAGS += -D__FILENAME__=\"$(notdir $<)\"
```

Comment:
I think you may need to tack this onto the rule directly, otherwise this would get expanded before you want?

Reply:
Don't believe so, as it works. But I think I should for clarity as well.

Reply:
indeed it does, neat!